nservicebus/messaging/evolving-contracts.md

```diff
@@ -1,108 +1,177 @@
+
+Ship multiple versions of a message contract in the same assembly by creating a new message type for the new contract version. The new type typically indicates its relationship via its name, e.g.
+
+```
+// first version of the message contract
+public class CreateOrder : ICommand {
+    ...
+}
+
+// new version of the message contract
+public class CreateOrderV2 : Icommand{
+    ...
+}
+```
+
+* Consumers can be updated to the new contract version independent from updating the contract assembly, giving more flexibility in planning the upgrade process.
+* Allows more flexibility about the contract changes as no type-level compatibility has to be ensured.
+
+When updating commands/messages:
+* Ensure the receiver of the message has a handler for both types to ensure it can process messages that are still in the queue or coming from endpoints still using the old contract.
+
+When updating events:
+* Update all subscribers to have handlers (and subscribe) to both events before changing the publisher.
+
+### Use inheritance
+
+Ship multiple versions of a message contract in the same assembly by using inheritance. The new contract version inherits from the previous version, gaining all its parent's properties.
+
+```
+// first version of the message contract
+public class OrderCreatedEvent : IEvent {
+    ...
+}
+
+public class OrderCreatedEventV2 : OrderCreatedEvent {
+    ...
+}
+```
+
+* Can only be used to make additive changes
+* Flattening the inheritance chain / removing older versions is more difficult compared to the other options
+* Something about multiple/all message handlers being invoked
+* Something about routing complexity?
+
+When updating commands/messages:
+* Update all senders to the new version before adding a new message handler for the new version. This might delay the time till the new message contract can be implemented by the receiver significantly.
+
+When updating events:
+* Update the publisher first to publish only the latest version of the event. Subscribers using the old message contract will also receive the event but process it the same as the previous version of the contract.
+
+### Modify contract (adding data)
+
+When modifying an existing contract, contract consumers updating to the latest contract version will only know this version of the contract. When only making additive changes, the serialization behavior can be used to ensure backwards compatibility with older versions of the contract.
+
+```
+// first version of the message contract
+public class CreateOrder : ICommand {
+    ...
+    // this property has been added in the latest version:
+    public int? SomeProperty { get; set; }
+}
+```
+
+* Endpoints using the old version of the contract can process newer versions since the additional data is just being ignored
+* The new data should be using nullable types to ensure endpoints can distinguish between missing data and intentionally set default values (e.g., `0` when using `int`).
```

Comment:
I have a feeling this is a mess to handle in reality. It's like when you have checks in the code to determine the meaning of a SQL field based on its nullability.